mongo/document/document.go

```diff
@@ -46,64 +48,85 @@ type Closer interface {
+	// It is not useful to have both defined on a document, only ValidationWithInternalError
+	// would be used in this case
+	Validate(ctx context.Context) *errors.ValidationErrors
+
+	// ValidateWithInternalError will be used in priority if defined on a document
+	// It will be called for all modifying operations (Create, Save, Update)
+	// If it returns an internal error, the validation error will be nil.
+	ValidateWithInternalError(ctx context.Context) (*errors.ValidationErrors, error)
 }
 
 var _ Validable = &Base{}
 
 // Create inserts the document in the database, returns an error if document
 // already exists and set CreatedAt timestamp
 func Create(ctx context.Context, collectionName string, doc document) error {
-	log := logger.Get(ctx).WithFields(logrus.Fields{
-		"collection": collectionName,
-		"doc_id":     doc.getID().Hex(),
+	return save(ctx, collectionName, doc, func(ctx context.Context, collectionName string, doc document) error {
+		log := logger.Get(ctx)
+		c := mongo.Session(log).Clone().DB("").C(collectionName)
+		defer c.Database.Session.Close()
+		log.WithFields(logrus.Fields{
+			"collection": collectionName,
+			"doc_id":     doc.getID().Hex(),
+		}).Debugf("save '%v'", collectionName)
+		return c.Insert(doc)
 	})
-	doc.ensureID()
-	doc.ensureCreatedAt()
-	doc.setUpdatedAt(time.Now())
+}
 
-	validationErrors, err := doc.Validate(ctx)
-	if err != nil {
-		log.WithError(err).Error("Internal error while validating the document")
+func Save(ctx context.Context, collectionName string, doc document) error {
+	return save(ctx, collectionName, doc, func(ctx context.Context, collectionName string, doc document) error {
+		log := logger.Get(ctx)
+		c := mongo.Session(log).Clone().DB("").C(collectionName)
+		defer c.Database.Session.Close()
+		log.WithFields(logrus.Fields{
+			"collection": collectionName,
+			"doc_id":     doc.getID().Hex(),
+		}).Debugf("save '%v'", collectionName)
+		_, err := c.UpsertId(doc.getID(), doc)
 		return err
+	})
+}
+
+func Update(ctx context.Context, collectionName string, update bson.M, doc document) error {
+	return save(ctx, collectionName, doc, func(ctx context.Context, collectionName string, doc document) error {
+		log := logger.Get(ctx)
+		c := mongo.Session(log).Clone().DB("").C(collectionName)
+		defer c.Database.Session.Close()
+
+		if _, ok := update["$set"]; ok {
+			update["$set"].(bson.M)["updated_at"] = doc.getUpdatedAt()
+		}
+
+		log.WithFields(logrus.Fields{
+			"collection": collectionName,
+			"doc_id":     doc.getID().Hex(),
+		}).Debugf("update %v", collectionName)
+		return c.UpdateId(doc.getID(), update)
+	})
+}
+
+func save(ctx context.Context, collectionName string, doc document, saveFunc func(context.Context, string, document) error) error {
+	validationErrors, err := doc.ValidateWithInternalError(ctx)
+	if err == ErrValidateNoInternalErrorFunc {
+		validationErrors = doc.Validate(ctx)
+	} else if err != nil {
+		return errors.Wrap(ctx, err, "fail to validate document")
```

Comment:
nitpick(non-blocking):

```suggestion
		return errors.Wrap(ctx, err, "validate document")
```

Reply:
Ack